File: src/test/java/cube8540/oauth/authentication/oauth/security/endpoint/DefaultOAuth2AccessTokenIntrospectionConverterTest.java
package cube8540.oauth.authentication.oauth.security.endpoint;

import cube8540.oauth.authentication.oauth.AccessTokenIntrospectionKey;
import cube8540.oauth.authentication.oauth.security.OAuth2AccessTokenDetails;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static cube8540.oauth.authentication.AuthenticationApplication.DEFAULT_TIME_ZONE;
import static cube8540.oauth.authentication.oauth.security.endpoint.TokenEndpointTestHelper.EXPIRATION;
import static cube8540.oauth.authentication.oauth.security.endpoint.TokenEndpointTestHelper.RAW_CLIENT_ID;
import static cube8540.oauth.authentication.oauth.security.endpoint.TokenEndpointTestHelper.RAW_SCOPES;
import static cube8540.oauth.authentication.oauth.security.endpoint.TokenEndpointTestHelper.RAW_USERNAME;
import static cube8540.oauth.authentication.oauth.security.endpoint.TokenEndpointTestHelper.makeAccessTokenDetails;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.mockito.Mockito.when;

@DisplayName("기본 엑세스 토큰")
class DefaultOAuth2AccessTokenIntrospectionConverterTest {

    @Test
    @DisplayName("액세스 토큰의 유저 명이 null 일때")
    void whenAccessTokenUsernameIsNull() {
        OAuth2AccessTokenDetails accessToken = makeAccessTokenDetails();
        DefaultOAuth2AccessTokenIntrospectionConverter converter = new DefaultOAuth2AccessTokenIntrospectionConverter();

        when(accessToken.getUsername()).thenReturn(null);

        Map<String, Object> result = converter.convertAccessToken(accessToken);
        assertEquals(RAW_CLIENT_ID, result.get(AccessTokenIntrospectionKey.CLIENT_ID));
        assertEquals(EXPIRATION.atZone(DEFAULT_TIME_ZONE.toZoneId()).toEpochSecond(), result.get(AccessTokenIntrospectionKey.EXPIRATION));
        assertEquals(String.join(" ", RAW_SCOPES), result.get(AccessTokenIntrospectionKey.SCOPE));
        assertNull(result.get(AccessTokenIntrospectionKey.USERNAME));
    }

    @Test
    @DisplayName("액세스 토큰이 만료 되었을 때")
    void whenAccessTokenExpiration() {
        OAuth2AccessTokenDetails accessToken = makeAccessTokenDetails();
        DefaultOAuth2AccessTokenIntrospectionConverter converter = new DefaultOAuth2AccessTokenIntrospectionConverter();

        when(accessToken.getExpired()).thenReturn(true);

        Map<String, Object> result = converter.convertAccessToken(accessToken);
        assertEquals(RAW_CLIENT_ID, result.get(AccessTokenIntrospectionKey.CLIENT_ID));
        assertEquals(EXPIRATION.atZone(DEFAULT_TIME_ZONE.toZoneId()).toEpochSecond(), result.get(AccessTokenIntrospectionKey.EXPIRATION));
        assertEquals(String.join(" ", RAW_SCOPES), result.get(AccessTokenIntrospectionKey.SCOPE));
        assertEquals(RAW_USERNAME, result.get(AccessTokenIntrospectionKey.USERNAME));
        assertFalse(Boolean.parseBoolean(result.get(AccessTokenIntrospectionKey.ACTIVE).toString()));
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/security/endpoint/DefaultOAuth2AccessTokenIntrospectionConverterTest.kt
package cube8540.oauth.authentication.oauth.security.endpoint

import cube8540.oauth.authentication.oauth.AccessTokenIntrospectionKey
import cube8540.oauth.authentication.oauth.security.OAuth2AccessTokenDetails
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

class DefaultOAuth2AccessTokenIntrospectionConverterTest {

    private val converter = DefaultOAuth2AccessTokenIntrospectionConverter()

    @Test
    fun `access token convert when username is null`() {
        val accessToken: OAuth2AccessTokenDetails = mockk {
            every { tokenValue } returns "tokenId"
            every { clientId } returns "clientId"
            every { username } returns null
            every { expiration } returns LocalDateTime.of(2020, 2, 1, 22, 52)
            every { scopes } returns setOf("scope-1", "scope-2", "scope-3")
            every { expired } returns false
        }

        val result = converter.convertAccessToken(accessToken)
        assertThat(result[AccessTokenIntrospectionKey.CLIENT_ID]).isNull()
        assertThat(result[AccessTokenIntrospectionKey.EXPIRATION]).isNull()
        assertThat(result[AccessTokenIntrospectionKey.SCOPE]).isEqualTo("scope-1 scope-2 scope-3")
        assertThat(result[AccessTokenIntrospectionKey.USERNAME]).isNull()
    }

    @Test
    fun `access token convert when token is expired`() {
        val accessToken: OAuth2AccessTokenDetails = mockk {
            every { tokenValue } returns "tokenId"
            every { clientId } returns "clientId"
            every { username } returns "username"
            every { expiration } returns LocalDateTime.of(2020, 2, 1, 22, 52)
            every { scopes } returns setOf("scope-1", "scope-2", "scope-3")
            every { expired } returns true
        }

        val result = converter.convertAccessToken(accessToken)
        assertThat(result[AccessTokenIntrospectionKey.CLIENT_ID]).isNull()
        assertThat(result[AccessTokenIntrospectionKey.EXPIRATION]).isNull()
        assertThat(result[AccessTokenIntrospectionKey.SCOPE]).isEqualTo("scope-1 scope-2 scope-3")
        assertThat(result[AccessTokenIntrospectionKey.USERNAME]).isNull()
        assertThat(result[AccessTokenIntrospectionKey.ACTIVE].toString().toBoolean()).isFalse
    }
}